# Banking Analytics Capstone Project

## Data Understanding

### Objective

The objective of this notebook is to understand the banking dataset before performing any cleaning or analysis.

In this phase, we will:

- Load the datasets
- Understand the dataset structure
- Check data types
- Validate primary and foreign keys
- Identify missing values
- Identify duplicate records
- Summarize data quality issues

No data will be modified in this notebook.

## 1. Import Libraries

In [19]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)

## 2. Load Dataset

In [20]:
customers = pd.read_csv("data/Raw/dim_customers.csv")
accounts = pd.read_csv("data/Raw/dim_accounts.csv")
cards = pd.read_csv("data/Raw/dim_cards.csv")
branches = pd.read_csv("data/Raw/dim_branches.csv")
merchants = pd.read_csv("data/Raw/dim_merchants.csv")
dates = pd.read_csv("data/Raw/dim_date.csv")
transactions = pd.read_csv("data/Raw/fact_transactions.csv")
loans = pd.read_csv("data/Raw/fact_loans.csv")
complaints = pd.read_csv("data/Raw/fact_complaints.csv")

## 3. Store Datasets

In [21]:
datasets = {
    "Customers": customers,
    "Accounts": accounts,
    "Cards": cards,
    "Branches": branches,
    "Merchants": merchants,
    "Date": dates,
    "Transactions": transactions,
    "Loans": loans,
    "Complaints": complaints
}

In [22]:
import pandas as pd

files = {
    "dim_customers": "data/Cleaned/dim_customers.csv",
    "dim_accounts": "data/Cleaned/dim_accounts.csv",
    "dim_cards": "data/Cleaned/dim_cards.csv",
    "dim_branches": "data/Cleaned/dim_branches.csv",
    "dim_merchants": "data/Cleaned/dim_merchants.csv",
    "dim_date": "data/Cleaned/dim_date.csv",
    "fact_transactions": "data/Cleaned/fact_transactions.csv",
    "fact_loans": "data/Cleaned/fact_loans.csv",
    "fact_complaints": "data/Cleaned/fact_complaints.csv"
}

for name, path in files.items():
    df = pd.read_csv(path, nrows=1)
    print(f"\n{'='*20} {name} {'='*20}")
    print(df.columns.tolist())


==================== dim_customers ====================
['customer_id', 'name', 'age', 'gender', 'occupation', 'employment_type', 'education', 'marital_status', 'income_bracket', 'monthly_income', 'credit_score', 'risk_category', 'customer_segment', 'kyc_status', 'signup_date', 'city', 'state', 'home_branch_id', 'is_digital_banking_user', 'is_upi_user', 'is_mobile_app_user', 'sms_alerts_enabled', 'is_insurance_holder']

==================== dim_accounts ====================
['account_id', 'customer_id', 'account_type', 'open_date', 'branch_id', 'status', 'current_balance', 'avg_monthly_balance', 'min_balance', 'interest_earned']

==================== dim_cards ====================
['card_id', 'account_id', 'card_type', 'issue_date', 'card_status', 'annual_fee', 'fee_waived']

==================== dim_branches ====================
['branch_id', 'branch_name', 'city', 'state', 'region', 'zone', 'latitude', 'longitude', 'manager_name', 'employee_count', 'atm_count', 'opening_year']

====

## 4. Dataset Overview

In [23]:
overview = pd.DataFrame({
    "Table": datasets.keys(),
    "Rows": [df.shape[0] for df in datasets.values()],
    "Columns": [df.shape[1] for df in datasets.values()]
})

overview

,Table,Rows,Columns
0,Customers,150000,23
1,Accounts,240000,10
2,Cards,300000,7
3,Branches,25,12
4,Merchants,3500,5
5,Date,1096,13
6,Transactions,2522773,13
7,Loans,120000,12
8,Complaints,11124,10


## 5. View Sample Records

In [24]:
customers.head()

,customer_id,name,age,gender,occupation,employment_type,education,marital_status,income_bracket,monthly_income,credit_score,risk_category,customer_segment,kyc_status,signup_date,city,state,home_branch_id,is_digital_banking_user,is_upi_user,is_mobile_app_user,sms_alerts_enabled,is_insurance_holder
0,CUST000001,Robert Garcia,18,Female,Student,Part-Time,Bachelor's,Divorced,Low,21373,523,High Risk,New,Verified,2024-12-03,Hyderabad,Telangana,BR0007,True,True,False,True,False
1,CUST000002,Janice Petty,65,Female,Salaried - Private,Part-Time,Diploma,Married,Upper-Middle,145501,834,Low Risk,Premium,Pending,2022-09-27,Hyderabad,Telangana,BR0007,True,False,True,False,True
2,CUST000003,Victor Waller,33,Male,Self-Employed,Contract,Bachelor's,Married,Upper-Middle,145831,828,Low Risk,Premium,Verified,2022-02-11,Chandigarh,Punjab,BR0002,True,True,True,True,False
3,CUST000004,Brenda Morrison,40,Male,Student,Self-Employed,Master's,Married,Lower-Middle,28208,669,Medium Risk,Regular,Verified,2023-11-08,Patna,Bihar,BR0014,True,False,True,True,True
4,CUST000005,Michael Valencia,21,Male,Salaried - Private,Self-Employed,Bachelor's,Single,Upper-Middle,125489,805,Low Risk,Premium,Verified,2024-05-11,Surat,Gujarat,BR0012,True,True,False,False,False


In [25]:
transactions.head()

,transaction_id,date_id,account_id,card_id,merchant_id,amount,transaction_type,channel,city,status,is_fraud,mdr_revenue,transaction_fee
0,SAL0289407,20220101,ACC033291,NaN,MER02633,34509.99,Deposit,Bank Transfer,Kolkata,Completed,False,0.0,0.0
1,SAL0563049,20220101,ACC064646,NaN,MER02081,29056.48,Deposit,Bank Transfer,Jaipur,Completed,False,0.0,0.0
2,TXN0379705,20220101,ACC215134,CRD244383,MER00568,1223.20,Purchase,ATM,Hyderabad,Completed,False,0.0,0.0
3,TXN0095155,20220101,ACC024154,CRD252211,MER02069,27858.24,Withdrawal,POS,Kolkata,Completed,False,0.0,0.0
4,SAL0815040,20220101,ACC094046,NaN,MER02142,25644.37,Deposit,Bank Transfer,Jaipur,Completed,False,0.0,0.0


## 6. Check Data Types

In [26]:
for name, df in datasets.items():
    print(f"\n{name}")
    print(df.dtypes)


Customers
customer_id                  str
name                         str
age                        int64
gender                       str
occupation                   str
employment_type              str
education                    str
marital_status               str
income_bracket               str
monthly_income             int64
credit_score               int64
risk_category                str
customer_segment             str
kyc_status                   str
signup_date                  str
city                         str
state                        str
home_branch_id               str
is_digital_banking_user     bool
is_upi_user                 bool
is_mobile_app_user          bool
sms_alerts_enabled          bool
is_insurance_holder         bool
dtype: object

Accounts
account_id                 str
customer_id                str
account_type               str
open_date                  str
branch_id                  str
status                     str
current_balance     

## 7. Convert Date Columns

In [27]:
customers["signup_date"] = pd.to_datetime(customers["signup_date"])
accounts["open_date"] = pd.to_datetime(accounts["open_date"])
cards["issue_date"] = pd.to_datetime(cards["issue_date"])
loans["disbursement_date"] = pd.to_datetime(loans["disbursement_date"])
dates["full_date"] = pd.to_datetime(dates["full_date"])

C:\Users\deepa\AppData\Local\Temp\ipykernel_6980\1813873562.py:2: UserWarning: Parsing dates in %d-%m-%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  accounts["open_date"] = pd.to_datetime(accounts["open_date"])


## 8. Dataset Information

In [28]:
customers.info()

<class 'pandas.DataFrame'>
RangeIndex: 150000 entries, 0 to 149999
Data columns (total 23 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   customer_id              150000 non-null  str           
 1   name                     150000 non-null  str           
 2   age                      150000 non-null  int64         
 3   gender                   147000 non-null  str           
 4   occupation               147000 non-null  str           
 5   employment_type          150000 non-null  str           
 6   education                147750 non-null  str           
 7   marital_status           150000 non-null  str           
 8   income_bracket           145500 non-null  str           
 9   monthly_income           150000 non-null  int64         
 10  credit_score             150000 non-null  int64         
 11  risk_category            150000 non-null  str           
 12  customer_segment         15

In [29]:
transactions.info()

<class 'pandas.DataFrame'>
RangeIndex: 2522773 entries, 0 to 2522772
Data columns (total 13 columns):
 #   Column            Dtype  
---  ------            -----  
 0   transaction_id    str    
 1   date_id           int64  
 2   account_id        str    
 3   card_id           str    
 4   merchant_id       str    
 5   amount            float64
 6   transaction_type  str    
 7   channel           str    
 8   city              str    
 9   status            str    
 10  is_fraud          bool   
 11  mdr_revenue       float64
 12  transaction_fee   float64
dtypes: bool(1), float64(3), int64(1), str(8)
memory usage: 233.4 MB


## 9. Missing Values

In [30]:
missing_summary = []

for name, df in datasets.items():
    missing = df.isnull().sum()
    missing = missing[missing > 0]
    for column, value in missing.items():
        missing_summary.append({
            "Table": name,
            "Column": column,
            "Missing Values": value,
            "Missing %": round(value/len(df)*100, 2)
        })

missing_summary = pd.DataFrame(missing_summary)
missing_summary

,Table,Column,Missing Values,Missing %
0,Customers,gender,3000,2.00
1,Customers,occupation,3000,2.00
2,Customers,education,2250,1.50
3,Customers,income_bracket,4500,3.00
4,Transactions,card_id,1310827,51.96
5,Transactions,merchant_id,37842,1.50
6,Transactions,channel,25228,1.00
7,Complaints,linked_transaction_id,9000,80.91
8,Complaints,resolution_days,3326,29.90


## 10. Duplicate Records

In [31]:
duplicate_summary = pd.DataFrame({
    "Table": datasets.keys(),
    "Duplicate Rows": [df.duplicated().sum() for df in datasets.values()]
})

duplicate_summary

,Table,Duplicate Rows
0,Customers,0
1,Accounts,0
2,Cards,0
3,Branches,0
4,Merchants,0
5,Date,0
6,Transactions,23372
7,Loans,0
8,Complaints,0


## 11. Primary Key Validation

In [32]:
pk_summary = pd.DataFrame({
    "Table": [
        "Customers", "Accounts", "Cards", "Branches", "Merchants",
        "Date", "Transactions", "Loans", "Complaints"
    ],
    "Duplicate Primary Keys": [
        customers["customer_id"].duplicated().sum(),
        accounts["account_id"].duplicated().sum(),
        cards["card_id"].duplicated().sum(),
        branches["branch_id"].duplicated().sum(),
        merchants["merchant_id"].duplicated().sum(),
        dates["date_id"].duplicated().sum(),
        transactions["transaction_id"].duplicated().sum(),
        loans["loan_id"].duplicated().sum(),
        complaints["complaint_id"].duplicated().sum()
    ]
})

pk_summary

,Table,Duplicate Primary Keys
0,Customers,0
1,Accounts,0
2,Cards,0
3,Branches,0
4,Merchants,0
5,Date,0
6,Transactions,24978
7,Loans,0
8,Complaints,0


## 12. Foreign Key Validation

In [33]:
relationships = [
    # Accounts
    ("Accounts → Customers", accounts, "customer_id", customers, "customer_id"),
    ("Accounts → Branches", accounts, "branch_id", branches, "branch_id"),

    # Cards
    ("Cards → Accounts", cards, "account_id", accounts, "account_id"),

    # Transactions
    ("Transactions → Accounts", transactions, "account_id", accounts, "account_id"),

    # Ignore NULL card_id values (Salary deposits, etc.)
    ("Transactions → Cards",
     transactions.dropna(subset=["card_id"]),
     "card_id",
     cards,
     "card_id"),

    # Ignore NULL merchant_id values (Internal transfers, salary credits, etc.)
    ("Transactions → Merchants",
     transactions.dropna(subset=["merchant_id"]),
     "merchant_id",
     merchants,
     "merchant_id"),

    ("Transactions → Date",
     transactions,
     "date_id",
     dates,
     "date_id"),

    # Loans
    ("Loans → Customers",
     loans,
     "customer_id",
     customers,
     "customer_id"),

    # Complaints
    ("Complaints → Customers",
     complaints,
     "customer_id",
     customers,
     "customer_id"),

    ("Complaints → Accounts",
     complaints,
     "account_id",
     accounts,
     "account_id"),

    # Ignore complaints that are not linked to a transaction
    ("Complaints → Transactions",
     complaints.dropna(subset=["linked_transaction_id"]),
     "linked_transaction_id",
     transactions,
     "transaction_id"),

    ("Complaints → Date",
     complaints,
     "date_id",
     dates,
     "date_id")
]



fk_results = []

for relation, child_df, child_col, parent_df, parent_col in relationships:

    invalid_records = child_df.loc[
        ~child_df[child_col].isin(parent_df[parent_col])
    ].shape[0]

    fk_results.append({
        "Relationship": relation,
        "Rows Checked": len(child_df),
        "Invalid Records": invalid_records,
        "Status": "✅ Pass" if invalid_records == 0 else "❌ Fail"
    })

fk_results = pd.DataFrame(fk_results)

fk_results

,Relationship,Rows Checked,Invalid Records,Status
0,Accounts → Customers,240000,0,✅ Pass
1,Accounts → Branches,240000,0,✅ Pass
2,Cards → Accounts,300000,0,✅ Pass
3,Transactions → Accounts,2522773,0,✅ Pass
4,Transactions → Cards,1211946,0,✅ Pass
5,Transactions → Merchants,2484931,0,✅ Pass
6,Transactions → Date,2522773,0,✅ Pass
7,Loans → Customers,120000,0,✅ Pass
8,Complaints → Customers,11124,0,✅ Pass
9,Complaints → Accounts,11124,0,✅ Pass


## 13. Basic Statistics

In [34]:
customers.describe()

,age,monthly_income,credit_score,signup_date
count,150000.000000,150000.000000,150000.000000,150000
mean,39.749800,80736.424227,682.863553,2022-07-02 22:21:08.352000
min,18.000000,12000.000000,349.000000,2020-01-02 00:00:00
25%,31.000000,32951.000000,618.000000,2021-03-31 00:00:00
50%,39.000000,56703.500000,680.000000,2022-07-05 00:00:00
75%,48.000000,97349.500000,747.000000,2023-10-02 00:00:00
max,85.000000,399999.000000,900.000000,2024-12-30 00:00:00
std,12.474393,76454.115137,90.997108,NaN


In [35]:
transactions.describe()

,date_id,amount,mdr_revenue,transaction_fee
count,2.522773e+06,2.522773e+06,2.522773e+06,2.522773e+06
mean,2.023277e+07,5.381816e+04,2.261492e+01,2.364914e-01
std,7.999719e+03,2.521448e+05,1.264428e+02,1.800454e+00
min,2.022010e+07,-1.970345e+07,0.000000e+00,0.000000e+00
25%,2.023011e+07,2.424800e+03,0.000000e+00,0.000000e+00
50%,2.023110e+07,2.356853e+04,0.000000e+00,0.000000e+00
75%,2.024070e+07,6.266944e+04,0.000000e+00,0.000000e+00
max,2.024123e+07,2.678984e+07,2.368502e+04,2.499000e+01


### Quick Outlier Check

Before moving to Cleaning, it's worth checking the extreme ends of the key numeric columns rather than just looking at the summary table above.

In [36]:
# Checking extreme values flagged by describe() above
print("Min transaction amount:", transactions["amount"].min())
print("Max transaction amount:", transactions["amount"].max())
print("Min customer age:", customers["age"].min())
print("Max customer age:", customers["age"].max())
print("Max monthly income:", customers["monthly_income"].max())

Min transaction amount: -19703454.213987708
Max transaction amount: 26789837.08949523
Min customer age: 18
Max customer age: 85
Max monthly income: 399999


## 14. Observations

- The project contains 9 related tables following a star schema (Customers, Accounts, Cards, Branches, Merchants, Date as dimensions; Transactions, Loans, Complaints as facts).
- All 12 foreign key relationships checked (Accounts, Cards, Transactions, Loans, Complaints against their parent tables) passed with 0 invalid records.
- No duplicate primary keys exist in any dimension table. Transactions has 24,978 duplicate `transaction_id` values, of which 23,372 are full-row duplicates and the remaining ~1,600 differ in at least one other column. This pattern is typical of duplicate ETL re-ingestion and will be handled in the Cleaning phase.
- Missing values fall into two groups: (1) genuinely optional fields such as `gender`, `occupation`, `education`, and `income_bracket` in Customers (1.5%-3% missing each), and (2) structurally explainable nulls such as `card_id` (51.96% missing — cash/UPI/internal transfers have no card) and `merchant_id` (1.5% missing — salary credits and internal transfers have no merchant). `linked_transaction_id` is missing for 80.91% of complaints, consistent with complaints that aren't tied to a specific transaction (e.g. service or KYC issues), and `resolution_days` is missing for 29.90% of complaints, consistent with complaints still open/unresolved.
- `transactions.amount` ranges from roughly -1.97 crore to +2.68 crore, and customer age/income also show wide ranges (age up to 85, monthly income up to ₹3,99,999). These extremes need to be investigated in the Cleaning phase to confirm whether they are genuine outliers, refunds/reversals, or data quality issues.
- Date columns (`signup_date`, `open_date`, `issue_date`, `disbursement_date`, `full_date`) were converted from string to datetime for future time-based analysis.
- The dataset structure is validated and ready for the Data Cleaning phase, with duplicate transactions and the amount/age/income outliers above as the main items to address first.